sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -470,15 +471,15 @@ object PartitioningUtils {
    * Given a collection of [[Literal]]s, resolves possible type conflicts by up-casting "lower"
    * types.
    */
-  private def resolveTypeConflicts(literals: Seq[Literal]): Seq[Literal] = {
+  private def resolveTypeConflicts(literals: Seq[Literal], timeZone: TimeZone): Seq[Literal] = {
     val desiredType = {
       val topType = literals.map(_.dataType).maxBy(upCastingOrder.indexOf(_))
```

Comment:
thanks for such a quick fix!

but, don't we also need to update [`upCastingOrder`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L430)?  It seems this happens to work because `upcastingOrder.indexOf(TimestampType) = -1`.  But I think that doesn't compare the right way with NullType, so if you add this to [`ParquetPartitionDiscoverySuite.test("parse partitions")`](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala#L241), it fails:

```
    check(Seq(
      s"hdfs://host:9000/path/a=$defaultPartitionName/b=blah",
      s"hdfs://host:9000/path/a=2014-01-01 00%3A00%3A00.0/b=foo"),
      PartitionSpec(
        StructType(Seq(
          StructField("a", TimestampType),
          StructField("b", StringType))),
        Seq(
          Partition(InternalRow(null, "blah"),
            s"hdfs://host:9000/path/a=$defaultPartitionName/b=blah"),
          Partition(InternalRow(Timestamp.valueOf("2014-01-01 00:00:00.0"), "foo"),
            s"hdfs://host:9000/path/a=2014-01-01 00%3A00%3A00.0/b=foo"))))
```

(I have to admit, I don't totally understand what the ramifications of that fail are -- the behavior in the resulting dataframe seems fine to me, but I figure there is probably some case this would mess up ...)

Reply:
I see. Actually, I believe this PR fixed only the (corner) regression between 2.1.0 and 2.2.0 because, if I understood correctly, we started to support infer timestamp in partition column from 2.1.0, [SPARK-17388](https://issues.apache.org/jira/browse/SPARK-17388), and this corner regression was introduced from 2.2.0 [SPARK-18939](https://issues.apache.org/jira/browse/SPARK-18939) (not tested).

For the issue described above, I think that also should be applied to `DecimalType`, `DateType` and `TimestampType` which were started to be supported from [SPARK-17388](https://issues.apache.org/jira/browse/SPARK-17388) and should strictly be related but orthogonal.

For perfectness, I guess we should port this logic:

https://github.com/apache/spark/blob/183d4cb71fbcbf484fc85d8621e1fe04cbbc8195/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L135-L150

because the problem is, `TimestampType`, `DateType` and `DecimalType` can be upcasted to other types easily by comparing numeric precedence, but of course with few special handling because we are currently only inferring decimals when `scale <= 0` (e.g., not 1.1) and castable to a decimal before trying a double:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L384-L393

This actually also could a problem of types between `DateType` and `TimestampType` which should be upcastable (from date to timestamp).

Let me take a closer look and probably make a fix soon.